src/compute/src/arrangement/merge_batcher_col.rs

```diff
@@ -0,0 +1,491 @@
+            pending: Vec::new(),
+        }
+    }
+
+    fn empty(&mut self) -> FixedStack<(D, T, R)> {
+        self.stash
+            .pop()
+            .unwrap_or_else(|| FixedStack::with_capacity(Self::buffer_size()))
+    }
+
+    /// Remove all elements from the stash.
+    fn clear_stash(&mut self) {
+        self.stash.clear();
+    }
+
+    /// Insert an empty buffer into the stash. Panics if the buffer is not empty.
+    fn recycle(&mut self, mut buffer: FixedStack<(D, T, R)>) {
+        if buffer.capacity() > 0 && self.stash.len() < 2 {
+            buffer.clear();
+            self.stash.push(buffer);
+        }
+    }
+
+    fn push(&mut self, batch: &mut Vec<(D, T, R)>) {
+        // Ensure `self.pending` has a capacity of `Self::buffer_size`.
+        if self.pending.capacity() < Self::buffer_size() {
+            self.pending
+                .reserve(Self::buffer_size() - self.pending.capacity());
+        }
+
+        while !batch.is_empty() {
+            self.pending.extend(
+                batch.drain(
+                    ..std::cmp::min(batch.len(), self.pending.capacity() - self.pending.len()),
+                ),
+            );
+            if self.pending.len() == self.pending.capacity() {
+                differential_dataflow::consolidation::consolidate_updates(&mut self.pending);
+                if self.pending.len() > self.pending.capacity() / 2 {
+                    // Flush if `self.pending` is more than half full after consolidation.
+                    self.flush_pending();
+                }
+            }
+        }
+    }
+
+    /// Move all elements in `pending` into `queue`. The data in `pending` must be compacted and
+    /// sorted. After this function returns, `self.pending` is empty.
+    fn flush_pending(&mut self) {
+        if !self.pending.is_empty() {
+            let mut stack = self.empty();
+            stack.reserve_items(self.pending.iter());
+            for tuple in self.pending.drain(..) {
+                stack.copy(&tuple);
+            }
+            self.queue_push(vec![stack]);
+            while self.queue.len() > 1
+                && (self.queue[self.queue.len() - 1].len()
+                    >= self.queue[self.queue.len() - 2].len() / 2)
+            {
+                let list1 = self.queue_pop().unwrap();
+                let list2 = self.queue_pop().unwrap();
+                let merged = self.merge_by(list1, list2);
+                self.queue_push(merged);
+            }
+        }
+    }
+
+    // This is awkward, because it isn't a power-of-two length any more, and we don't want
+    // to break it down to be so.
+    fn push_list(&mut self, list: Vec<FixedStack<(D, T, R)>>) {
+        while self.queue.len() > 1 && self.queue[self.queue.len() - 1].len() < list.len() {
+            let list1 = self.queue_pop().unwrap();
+            let list2 = self.queue_pop().unwrap();
+            let merged = self.merge_by(list1, list2);
+            self.queue_push(merged);
```

Comment:
uncovered

Reply:
(Moved to #25196) This is actually interesting -- we call it from seal to return data that isn't due yet, but in this setting, we just drained everything that was enqueued, so the queue is empty. This API is copied from Differential, which exposes it, and needs to handle the situation where someone outside adds a new list to the merge batcher.

I think we can change the implementation to not offer the API, but it's a departure of DD's interface.